src/project.rs

```diff
@@ -0,0 +1,95 @@
+use std::error::Error;
+use std::ffi::OsString;
+use std::fs::{read_dir, DirEntry};
+use std::io::prelude::*;
+
+pub type ProjectError = String;
+pub type Proc<T> = Result<T, ProjectError>;
+
+pub fn discover(dir: &str) -> Proc<Vec<GitRepo>> {
+    match read_dir(dir) {
+        Err(error) => Err(error.description().to_string()),
+        Ok(files) => {
+            for file_result in files {
+                match file_result {
+                    Ok(file) => {
+                        let x = try!(git_repo(file));
+                        println!("Found path: {:?}", x);
+                    }
+                    Err(err) => {
+                        return Err(err.description().to_string());
+                    }
+                }
+            }
+            Err("Finished work".to_string())
+        },
+    }
+}
+
+// TODO: Better parameter type for representing files?
+pub fn git_repo(root: DirEntry) -> Proc<GitRepo> {
+    match read_dir(root.path()) {
+        Err(error) =>
+            Err(error.description().to_string()),
+        Ok(files) => {
+            let v: Vec<GitRepo> = files.filter_map(|f| {
+                f.ok().and_then(|file| {
+                    file.metadata().ok().and_then(|fd| {
```

Comment:
`.ok()` returns an `Option` of the successful `Result` case. `and_then` takes a function from that value (call it `A`) to an `Option<B>` and returns an `Option<B>` by flattening the `None` cases into `None`. This is a great way to avoid `panic` from `filter_map` without having to `match` all the `Err` cases.